platform-hub-api/app/services/kubernetes/token_claim_service.rb

```diff
@@ -0,0 +1,75 @@
+module Kubernetes
+  module TokenClaimService
+    extend TokenLookup
+    extend self
+
+    module Errors
+      class TokenNotFound < StandardError; end
+      class TokenAlreadyDefinedForCluster < StandardError; end
+      class TokenInvalid < StandardError; end
+      class TokenAlreadyClaimed < StandardError; end
+    end
+
+    def claim_token(user, token)
+      claim = lookup(token)
+
+      if claim.nil?
+        raise Errors::TokenNotFound, "Token `#{token}` not found!"
+      end
+
+      if claim.user.nil?
+        begin
+          migrate_claimed_token_to_user_kubernetes_identity(user, claim)
+          remove_claimed_token_from_static_list(claim)
+        rescue => e
+          raise
+        end
+        [ claim.cluster, "Claimed `#{claim.cluster}` token." ]
+
+      elsif claim.user.present?
+        raise Errors::TokenAlreadyClaimed, "Token already claimed!"
+      end
+    end
+
+    private
+
+    def migrate_claimed_token_to_user_kubernetes_identity(user, claim)
+      identity = 
+        user.identity(:kubernetes) || 
+        user.identities.create!(
+          provider: :kubernetes,
+          external_id: user.email,
+          data: {tokens: []}
+        )
+
+      if identity.data['tokens'].find {|t| t['cluster'] == claim.cluster}
+        raise Errors::TokenAlreadyDefinedForCluster, 
+              "User `#{user.email}` already have a token assigned for `#{claim.cluster}` cluster!"
```

Comment:
I think that's actually quite possible scenario. Agree with your suggestion here. Let's allow user claiming their old token to override exiting one (if present) - in such case we should ensure that groups allocation allocated by the admin (in step 2 above) takes precedence over the groups in the old token.